Internal Plugin Code Review PR

dbluhm opened the pull request.

burdettadam committed:
feat: first of oid4vci plugin files

Signed-off-by: Adam Burdett <burdettadam@gmail.com>

burdettadam committed:
BUG: non working docker

Signed-off-by: Adam Burdett <burdettadam@gmail.com>

burdettadam committed:
BUG: non working docker

Signed-off-by: Adam Burdett <burdettadam@gmail.com>

burdettadam committed:
feat: working docker img

Signed-off-by: Adam Burdett <burdettadam@gmail.com>

burdettadam committed:
feat: loading plugin with status points

Signed-off-by: Adam Burdett <burdettadam@gmail.com>

burdettadam committed:
feat: empty handlers

Signed-off-by: Adam Burdett <burdettadam@gmail.com>

cjhowland committed:
feat: add logic to get_cred_offer endpoint

Signed-off-by: Char Howland <char@indicio.tech>

cjhowland committed:
feat: create CredentialOfferRecord

Signed-off-by: Char Howland <char@indicio.tech>

burdettadam committed:
feat: some openapi schemas

Signed-off-by: Adam Burdett <burdettadam@gmail.com>

burdettadam committed:
feat: no head allowed on get's

Signed-off-by: Adam Burdett <burdettadam@gmail.com>

cjhowland committed:
feat: oid_cred_issuer logic (WIP)

Signed-off-by: Char Howland <char@indicio.tech>

burdettadam committed:
feat: check jwt token

Signed-off-by: Adam Burdett <burdettadam@gmail.com>

burdettadam committed:
feat: mostly copied routes.py

Signed-off-by: Adam Burdett <burdettadam@gmail.com>

burdettadam committed:
feat: demo directory

Signed-off-by: Adam Burdett <burdettadam@gmail.com>

burdettadam committed:
feat: demo frontend with docker compose

Signed-off-by: Adam Burdett <burdettadam@gmail.com>

burdettadam committed:
feat: demo docker compose with less bugs

Signed-off-by: Adam Burdett <burdettadam@gmail.com>

burdettadam committed:
feat: cred ex record  management

Signed-off-by: Adam Burdett <burdettadam@gmail.com>

dbluhm committed:
chore: add pre-commit

Signed-off-by: Daniel Bluhm <dbluhm@pm.me>

dbluhm committed:
refactor: remove unused record

Signed-off-by: Daniel Bluhm <dbluhm@pm.me>

dbluhm committed:
refactor: models into model package

Signed-off-by: Daniel Bluhm <dbluhm@pm.me>

dbluhm committed:
refactor: minor tweaks to issuer meta endpoint

Signed-off-by: Daniel Bluhm <dbluhm@pm.me>

dbluhm committed:
refactor: move public routes to new module

Signed-off-by: Daniel Bluhm <dbluhm@pm.me>

burdettadam committed:
feat: more code

Signed-off-by: Adam Burdett <burdettadam@gmail.com>

dbluhm committed:
refactor: minor tweaks, draft 13 compliance

Signed-off-by: Daniel Bluhm <dbluhm@pm.me>

burdettadam committed:
feat: docker_compose with ngrok

Signed-off-by: Adam Burdett <burdettadam@gmail.com>

dbluhm committed:
fix: exchange record schema issues

Signed-off-by: Daniel Bluhm <dbluhm@pm.me>

dbluhm committed:
refactor: shorten the name of exchange record

Signed-off-by: Daniel Bluhm <dbluhm@pm.me>

dbluhm committed:
refactor: rename model module

Signed-off-by: Daniel Bluhm <dbluhm@pm.me>

dbluhm committed:
refactor: rename records a bit

Signed-off-by: Daniel Bluhm <dbluhm@pm.me>

dbluhm committed:
refactor: rename model module

Signed-off-by: Daniel Bluhm <dbluhm@pm.me>

dbluhm committed:
refactor: adjust naming of record id

Signed-off-by: Daniel Bluhm <dbluhm@pm.me>

burdettadam committed:
feat: less bugs

Signed-off-by: Adam Burdett <burdettadam@gmail.com>

dbluhm committed:
fix: record value after save

Signed-off-by: Daniel Bluhm <dbluhm@pm.me>

dbluhm committed:
feat: add initial framework for testing against afj

Signed-off-by: Daniel Bluhm <dbluhm@pm.me>

dbluhm committed:
feat: message buffering and TCP

Signed-off-by: Daniel Bluhm <dbluhm@pm.me>

dbluhm committed:
refactor: rearranging

Signed-off-by: Daniel Bluhm <dbluhm@pm.me>

dbluhm committed:
feat: add docker-compose setup and test

Signed-off-by: Daniel Bluhm <dbluhm@pm.me>

dbluhm committed:
feat: separate pieces into reusable chunks

Signed-off-by: Daniel Bluhm <dbluhm@pm.me>

dbluhm committed:
chore: add jsdocs

Signed-off-by: Daniel Bluhm <dbluhm@pm.me>

dbluhm committed:
chore: use remote json-rpc-api-proxy lib

Signed-off-by: Daniel Bluhm <dbluhm@pm.me>

burdettadam committed:
feat: more bugs with mt tests

Signed-off-by: Adam Burdett <burdettadam@gmail.com>

burdettadam committed:
feat: changes to records

Signed-off-by: Adam Burdett <burdettadam@gmail.com>

burdettadam committed:
feat: broken token issueance

Signed-off-by: Adam Burdett <burdettadam@gmail.com>

dbluhm committed:
refactor: code review and draft 11 vs 13

Signed-off-by: Daniel Bluhm <dbluhm@pm.me>

burdettadam committed:
feat: issue token untested handler

Signed-off-by: Adam Burdett <burdettadam@gmail.com>

burdettadam committed:
feat: multiple id bug fix

Signed-off-by: Adam Burdett <burdettadam@gmail.com>

burdettadam committed:
feat: meta sup id bug

Signed-off-by: Adam Burdett <burdettadam@gmail.com>

dbluhm committed:
fix: draft 11 metadata structure

Signed-off-by: Daniel Bluhm <dbluhm@pm.me>

dbluhm committed:
fix: extract identifier from cred sup create

Signed-off-by: Daniel Bluhm <dbluhm@pm.me>

dbluhm committed:
test: add tests for routes

Signed-off-by: Daniel Bluhm <dbluhm@pm.me>

dbluhm committed:
fix: display is a list of objects

Signed-off-by: Daniel Bluhm <dbluhm@pm.me>

burdettadam committed:
fix: token endpoint

Signed-off-by: Adam Burdett <burdettadam@gmail.com>

burdettadam committed:
feat: token endpoint issues jwt

Signed-off-by: Adam Burdett <burdettadam@gmail.com>

dbluhm committed:
fix: minor cleanup

Signed-off-by: Daniel Bluhm <dbluhm@pm.me>

dbluhm committed:
feat: integration test anticipated changes

Signed-off-by: Daniel Bluhm <dbluhm@pm.me>

dbluhm committed:
chore: remove sphereon int

Signed-off-by: Daniel Bluhm <dbluhm@pm.me>

dbluhm committed:
test: implement client for int tests

Signed-off-by: Daniel Bluhm <dbluhm@pm.me>

burdettadam committed:
feat: credential endpoint code

Signed-off-by: Adam Burdett <burdettadam@gmail.com>

burdettadam committed:
feat: more credential endpoint code

Signed-off-by: Adam Burdett <burdettadam@gmail.com>

dbluhm committed:
feat: more work on issuance

Signed-off-by: Daniel Bluhm <dbluhm@pm.me>

burdettadam committed:
feat: more credential endpoint code for front end

Signed-off-by: Adam Burdett <burdettadam@gmail.com>

burdettadam committed:
feat: more credential endpoint code for manual testing

Signed-off-by: Adam Burdett <burdettadam@gmail.com>

burdettadam committed:
feat: issue credential endpoint with app succuss

Signed-off-by: Adam Burdett <burdettadam@gmail.com>

dbluhm commented on configs/multitenant-plugins.yml:
Not sure how this got here...

burdettadam replied:
same. I will remove it.

dbluhm commented on oid4vci/.pre-commit-config.yaml:
We can probably bump the versions on these... Or delegate to a repo-wide pre-commit?

burdettadam replied:
I think we should do a repo-wide pre-commit.

dbluhm commented on oid4vci/README.md:
We should do a small write up of what's implemented

dbluhm commented on oid4vci/demo/docker-compose.yaml:
```diff
@@ -0,0 +1,95 @@
+      - NODE_ENV=development # Set NODE_ENV to 'development'
+      - WDS_SOCKET_PORT=0
+
+  cache:
+    image: redis:6.2-alpine
+    restart: always
+    ports:
+      - '6379:6379'
+    command: redis-server --save 20 1 --loglevel warning --requirepass eYVX7EwVmmxKPCDmwMtyKVge8oLd2t81
+    volumes:
+      - cache:/data
```
Is this used? I don't think it is; recommended for removal

burdettadam replied:
good catch!

dbluhm commented on oid4vci/demo/docker-compose.yaml:
```diff
@@ -0,0 +1,95 @@
+
+volumes:
+  cache:
+    driver: local
```
Along with these lines (see previous comment)

dbluhm commented on oid4vci/demo/frontend/README.md:
We should do a quick audit to check how much of this actually needs to be included with the demo frontend.

dbluhm commented on oid4vci/demo/frontend/src/FormPage.js:
```diff
@@ -0,0 +1,100 @@
+        console.log("get offer params:");
+        console.log(queryParams);
+        axios
+          .get("http://localhost:3001/oid4vci/draft-11/credential-offer", {
```
We should remove `draft-11` from the endpoint for now, I think

dbluhm commented on oid4vci/demo/frontend/src/FormPage.js:
```diff
@@ -0,0 +1,100 @@
+          height={150}
+          alt="UniversityDegreeCredential"
+          onClick={() => handleCredentialSelect("UniversityDegreeCredential")}
+        />
```
This might be confusing for people running the demo; should we only present a single credential?

dbluhm commented on oid4vci/int/oid4vci_client/client.py:
We should include a README to describe the purpose and goals of this client (i.e. it's meant for testing and that's about it)

dbluhm commented on oid4vci/int/tests/test_pre_auth_code_flow.py:
```diff
@@ -0,0 +1,57 @@
+            "id": "UniversityDegreeCredential",
+            "types": ["VerifiableCredential", "UniversityDegreeCredential"],
+        },
+    )
```
A demo setup file would basically just need to do this and set up a DID, I think

dbluhm commented on oid4vci/oid4vci/v1_0/__init__.py:
```diff
@@ -0,0 +1,49 @@
+        # -o oid4vci.host=0.0.0.0 -o oid4vci.port=8081
+        plugin_settings = profile.context.settings.for_plugin("oid4vci")
+        host = plugin_settings.get("host") or OID4VCI_HOST
+        port = plugin_settings.get("port") or OID4VCI_PORT
```
These need to be documented somewhere

dbluhm commented on oid4vci/oid4vci/v1_0/__init__.py:
```diff
@@ -0,0 +1,49 @@
+    """Setup the plugin."""
+    LOGGER.info("> oid4vci plugin setup...")
+    event_bus = context.inject(EventBus)
+    event_bus.subscribe(STARTUP_EVENT_PATTERN, started_event_handler)
```
We need a shutdown handler, too, that calls shutdown on our server

dbluhm commented on oid4vci/oid4vci/v1_0/models/exchange.py:
```diff
@@ -0,0 +1,119 @@
+    )
+    code = fields.Str(
+        required=False,
+    )
```
Light descriptions would be nice

dbluhm commented on oid4vci/oid4vci/v1_0/public_routes.py:
Commented out lines in this file should be cleaned up

dbluhm commented on oid4vci/oid4vci/v1_0/public_routes.py:
```diff
@@ -0,0 +1,328 @@
+
+@docs(tags=["oid4vci"], summary="Get credential issuer metadata")
+# @querystring_schema(TokenRequestSchema())
+async def oid_cred_issuer(request: web.Request):
```
I think it would be valuable to name these as closely as possible to what the spec calls them. For example, `credential_issuer_metadata`.

dbluhm commented on oid4vci/oid4vci/v1_0/routes.py:
Light suggestion, we should maybe consider following the routes/manager split pattern commonly used in ACA-Py. This can be something we worry about later but it would probably help extending this later.

dbluhm commented on oid4vci/oid4vci/v1_0/routes.py:
```diff
@@ -0,0 +1,468 @@
+    For example, can be used in QR-Code presented to a compliant wallet.
+    """
+    issuer_url = getenv("OID4VCI_ENDPOINT")
+    user_pin_required = getenv("OID4VCI_USER_PIN_REQUIRED", False)
```
These need to be documented somewhere (and pulled from plugin config value too, potentially)

dbluhm commented on oid4vci/oid4vci/v1_0/routes.py:
```diff
@@ -0,0 +1,468 @@
+async def register(app: web.Application):
+    """Register routes."""
+    # add in the message list(s) route
+    app.add_routes(
+        [
+            web.get(
+                "/oid4vci/draft-11/credential-offer", get_cred_offer, allow_head=False
+            ),
+            web.get(
+                "/oid4vci/draft-13/credential-offer", get_cred_offer, allow_head=False
+            ),
+            web.get(
+                "/oid4vci/exchange/records",
+                credential_exchange_list,
+                allow_head=False,
+            ),
+            web.post("/oid4vci/exchange/create", credential_exchange_create),
+            web.delete(
+                "/oid4vci/exchange/records/{cred_ex_id}",
+                credential_exchange_remove,
+            ),
+            web.post(
+                "/oid4vci/credential-supported/create", credential_supported_create
+            ),
+            web.get(
+                "/oid4vci/credential-supported/records",
+                credential_supported_list,
+                allow_head=False,
+            ),
+            web.delete(
+                "/oid4vci/exchange-supported/records/{cred_sup_id}",
+                credential_supported_list,
+            ),
+        ]
+    )
```
Let's drop the draft-11 vs 13 split for now.

dbluhm commented on oid4vci/setup.cfg:
I think this file can be removed

dbluhm commented on oid4vci/setup.py:
I think this file can be removed

dbluhm commented on oid4vci/tests/__init__.py:
We should add more tests eventually...

dbluhm reviewed (commented):
Thoughts:
- We should have a file that does setup for the demo on the ACA-Py side so when the user clicks through the frontend, the backend is ready to issue a credential of the type coming from the frontend.
- Should we leave off the AFJ testing for now? Put it in a separate branch to return to when the time is right?
- Same with the sphereon stuff?

burdettadam committed:
fix: removed some dead code

Signed-off-by: Adam Burdett <burdettadam@gmail.com>

dbluhm committed:
refactor: amended remove dead code

Signed-off-by: Daniel Bluhm <dbluhm@pm.me>

dbluhm committed:
feat: add config object

For reliably retrieving configuration values from plugin config or
environment

Signed-off-by: Daniel Bluhm <dbluhm@pm.me>

dbluhm committed:
refactor: rearrange plugin structure slightly

Since this is not a DIDComm protocol plugin, it doesn't make sense to
follow the same conventions

Signed-off-by: Daniel Bluhm <dbluhm@pm.me>

dbluhm committed:
refactor: refine configuration object

Signed-off-by: Daniel Bluhm <dbluhm@pm.me>

dbluhm committed:
refactor: further refine config object; errors

Signed-off-by: Daniel Bluhm <dbluhm@pm.me>

dbluhm committed:
refactor: minimize oid4vci server responsibilities

I think some of the previous endpoints are undesirable for public
endpoints.

Signed-off-by: Daniel Bluhm <dbluhm@pm.me>

dbluhm committed:
chore: small clean up in public routes

Signed-off-by: Daniel Bluhm <dbluhm@pm.me>

dbluhm committed:
refactor: minor rearrangement of route models

Signed-off-by: Daniel Bluhm <dbluhm@pm.me>

dbluhm committed:
refactor: refine record listing

Signed-off-by: Daniel Bluhm <dbluhm@pm.me>

dbluhm committed:
refactor: refine admin endpoints further

PIN should be set on record creation

Signed-off-by: Daniel Bluhm <dbluhm@pm.me>

dbluhm committed:
feat: implement supported credential deletion

Signed-off-by: Daniel Bluhm <dbluhm@pm.me>

dbluhm committed:
feat: implement pin support and token expiry

Signed-off-by: Daniel Bluhm <dbluhm@pm.me>

dbluhm committed:
feat: add did:jwk resolver

Signed-off-by: Daniel Bluhm <dbluhm@pm.me>

dbluhm committed:
feat: implement proof of possession verification

Signed-off-by: Daniel Bluhm <dbluhm@pm.me>

dbluhm committed:
fix: save record state on offer created

Signed-off-by: Daniel Bluhm <dbluhm@pm.me>

dbluhm committed:
fix: failing tests, missing config

Signed-off-by: Daniel Bluhm <dbluhm@pm.me>

dbluhm committed:
fix: types checking, issues in int tests

Signed-off-by: Daniel Bluhm <dbluhm@pm.me>

dbluhm committed:
fix: signature is b64url encoded

Signed-off-by: Daniel Bluhm <dbluhm@pm.me>

dbluhm committed:
test: es256k pop

Signed-off-by: Daniel Bluhm <dbluhm@pm.me>

dbluhm committed:
fix: small fixes

Signed-off-by: Daniel Bluhm <dbluhm@pm.me>

dbluhm committed:
chore: bump askar, acapy versions

Signed-off-by: Daniel Bluhm <dbluhm@pm.me>

burdettadam committed:
fix: add registration admin page

Signed-off-by: Adam Burdett <burdettadam@gmail.com>

burdettadam committed:
fix: remove credential page, fix get by id

Signed-off-by: Adam Burdett <burdettadam@gmail.com>

burdettadam committed:
fix: cleaner demo code with async

Signed-off-by: Adam Burdett <burdettadam@gmail.com>

burdettadam committed:
fix: vars names

Signed-off-by: Adam Burdett <burdettadam@gmail.com>

burdettadam committed:
feat: better error reporting

Signed-off-by: Adam Burdett <burdettadam@gmail.com>

burdettadam committed:
feat: webpage name change

Signed-off-by: Adam Burdett <burdettadam@gmail.com>

dbluhm committed:
feat: update readme

Signed-off-by: Daniel Bluhm <dbluhm@pm.me>
